File: app/src/main/java/com/klinker/android/twitter_l/data/roomdb/FavoriteTweet.java
package com.klinker.android.twitter_l.data.roomdb;


import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.PrimaryKey;
import twitter4j.Status;

@Entity(tableName = "favorite_tweets")
public class FavoriteTweet {

    @PrimaryKey
    public Long id;

    @ColumnInfo
    public int account;

    @ColumnInfo(name = "is_unread")
    public boolean isUnread;

    @ColumnInfo(name = "tweet_id")
    public long tweetId;

    public FavoriteTweet(long id, int account, boolean isUnread, long tweetId) {
        this.id = id;
        this.account = account;
        this.isUnread = isUnread;
        this.tweetId = tweetId;
    }

    public FavoriteTweet(Status status, int account) {

        this.id = null;
        this.account = account;
        this.isUnread = true;
        this.tweetId = status.getId();
    }

}

File: app/src/main/java/com/klinker/android/twitter_l/data/roomdb/FavoriteTweet.kt
package com.klinker.android.twitter_l.data.roomdb

import androidx.room.*
import twitter4j.Status


@Entity(tableName = "favorite_tweets", indices = [Index(value = ["tweet_id", "account"], unique = true)])
data class FavoriteTweet @JvmOverloads constructor(
        @Embedded val tweet: Tweet,
        val account: Int,
        @ColumnInfo(name = "is_unread") val isUnread: Boolean = false,
        @PrimaryKey(autoGenerate = true) val id: Long = 0
) : BaseDao.TalonEntity<FavoriteTweet> {
    @JvmOverloads
    constructor(status: Status, account: Int, isUnread: Boolean = false) : this(Tweet(status), account, isUnread)

    override fun copyWithId(id: Long): FavoriteTweet {
        return this.copy(id = id)
    }

}

@Dao
abstract class FavoriteTweetDao : BaseDao<FavoriteTweet>() {

    @Query("SELECT * FROM favorite_tweets WHERE account = :account ORDER BY tweet_id DESC LIMIT :pageSize OFFSET ((:page - 1) * :pageSize)")
    abstract fun getFavoriteTweets(account: Int, page: Int = 1, pageSize: Int = 200) : List<FavoriteTweet>

    @Query("DELETE FROM favorite_tweets WHERE account = :account AND id NOT IN (SELECT id FROM favorite_tweets WHERE account = :account ORDER BY id DESC LIMIT :size)")
    abstract fun trimDatabase(account: Int, size: Int)

    @Ignore
    fun create(status: Status, account: Int, isUnread: Boolean = false) : FavoriteTweet {
        return this.insert(FavoriteTweet(status, account, isUnread))
    }
}

